File: app/src/main/java/com/example/test/LoginActivity.java
package com.example.test;

import androidx.appcompat.app.AppCompatActivity;

import android.Manifest;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.telephony.TelephonyManager;
import android.text.Editable;
import android.text.TextWatcher;
import android.text.method.PasswordTransformationMethod;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import com.example.test.model.User;
import com.example.test.request.RequestLogin;
import com.example.test.result.ResultLogin;
import com.example.test.service.APIClient;
import com.example.test.service.APIInterface;
import com.example.test.utils.ProfileUtil;
import com.example.test.utils.ProgressDialogUtil;

import org.json.JSONObject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginActivity extends AppCompatActivity {

    private EditText input_username, input_password;
    private FrameLayout clearEmailLogin;
    private ImageButton show, hide;
    private APIInterface apiInterface;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        input_username = findViewById(R.id.input_username);
        input_password = findViewById(R.id.input_password);
        clearEmailLogin = findViewById(R.id.clearEmailLogin);
        show = findViewById(R.id.show);
        hide = findViewById(R.id.hide);

        input_username.setFocusable(true);



        findViewById(R.id.btnLogin).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                submitLogin();
            }
        });

        apiInterface      = APIClient.getClient(getApplicationContext()).create(APIInterface.class);


        User userPicker = new ProfileUtil().getUser(getApplicationContext());
        if (userPicker != null){
            Intent intent = new Intent(getApplicationContext(), MainActivity.class);
            startActivity(intent);
            finish();
        }
        initProperty();
    }

    private void submitLogin(){
        String email        = input_username.getText().toString().trim();
        String password    = input_password.getText().toString().trim();

        if (email.equals("")){
            input_username.setError("please input username");
            return;
        }
        if (password.equals("")){
            input_password.setError("please input password");
            return;
        }
        /*if (!email.equals("")){
            boolean validateMail = isEmailValid(email);
            if (!validateMail){
                input_username.setError("Format penulisan email tidak valid");
                return;
            }
        }*/



        /*Intent intent = new Intent(LoginActivity.this, MainActivity.class);
        startActivity(intent);*/


        try {
            final ProgressDialogUtil progressDialogUtil = new ProgressDialogUtil(LoginActivity.this, ProgressDialog.STYLE_SPINNER,false);
            progressDialogUtil.show();

            RequestLogin data = new RequestLogin(email, password);
            Call<ResultLogin> callLogin = apiInterface.doLogin(data);
            callLogin.enqueue(new Callback<ResultLogin>() {
                @Override
                public void onResponse(Call<ResultLogin> call, Response<ResultLogin> response) {
                    if(response.isSuccessful()){
                        String status = response.body().getStatus();
                        String message = response.body().getMessage();
                        if (status!=null && status.equals("ERR-00-000")){
//                            if (response.body().getBlock()){
                                User user = response.body().getData();
                                new ProfileUtil().save(getApplicationContext(), user);
                                Intent intent = new Intent(getApplicationContext(), MainActivity.class);
                                startActivity(intent);
                                finish();
//                            }else{
//                                Toast.makeText(LoginActivity.this,message,Toast.LENGTH_SHORT).show();
//                            }


                        }else{
                            Toast.makeText(LoginActivity.this,message,Toast.LENGTH_SHORT).show();
                        }


                    }else{
                        try {
                            JSONObject j = new JSONObject(response.errorBody().string());
                            Toast.makeText(LoginActivity.this,j.getString("message"),Toast.LENGTH_SHORT).show();
                        }catch (Exception e){
                            e.getMessage();
                        }
                    }
                    progressDialogUtil.dismiss();

                }

                @Override
                public void onFailure(Call<ResultLogin> call, Throwable t) {
                    Log.e("Retrofit Get", t.toString());
                    Toast.makeText(LoginActivity.this,"Invalid connection",Toast.LENGTH_SHORT).show();
                    progressDialogUtil.dismiss();
                }
            });
        }catch (Exception ex){
            ex.printStackTrace();
        }
    }



    private void initProperty(){
        input_username.addTextChangedListener(new TextWatcher() {
            public void afterTextChanged(Editable s) {
                if (input_username.getText().toString().length() == 0) {
                    input_username.setVisibility(View.VISIBLE);
                    clearEmailLogin.setVisibility(View.INVISIBLE);
                } else {
                    input_username.setError(null);
                }
            }
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                clearEmailLogin.setVisibility(View.VISIBLE);
            }
        });

        input_password.addTextChangedListener(new TextWatcher() {
            public void afterTextChanged(Editable s) {
                if (input_password.getText().toString().length() == 0) {
                    show.setVisibility(View.INVISIBLE);
                    hide.setVisibility(View.INVISIBLE);
                }else if(hide.getVisibility()==View.VISIBLE){
                    show.setVisibility(View.INVISIBLE);
                }else{
                    show.setVisibility(View.VISIBLE);
                }
            }

            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }
        });
        show.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                input_password.setTransformationMethod(null);
                show.setVisibility(View.INVISIBLE);
                hide.setVisibility(View.VISIBLE);
            }
        });
        hide.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                input_password.setTransformationMethod(new PasswordTransformationMethod());
                show.setVisibility(View.VISIBLE);
                hide.setVisibility(View.INVISIBLE);
            }
        });
        clearEmailLogin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                input_username.setText("");
                input_username.setVisibility(View.VISIBLE);
            }
        });

    }

    boolean doubleBackToExitPressedOnce = false;
    @Override
    public void onBackPressed() {
        //Checking for fragment count on backstack
        if (getSupportFragmentManager().getBackStackEntryCount() > 0) {
            getSupportFragmentManager().popBackStack();
        } else if (!doubleBackToExitPressedOnce) {
            this.doubleBackToExitPressedOnce = true;
            Toast.makeText(this,"Press BACK again to exit.", Toast.LENGTH_SHORT).show();

            new Handler().postDelayed(new Runnable() {

                @Override
                public void run() {
                    doubleBackToExitPressedOnce = false;
                }
            }, 2000);
        } else {
            super.onBackPressed();
            return;
        }
    }
}

File: app/src/main/java/com/example/test/LoginActivity.kt
package com.example.test

import android.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.ImageButton
import com.example.test.service.APIInterface
import android.os.Bundle
import com.example.test.R
import com.example.test.service.APIClient
import com.example.test.utils.ProfileUtil
import android.content.Intent
import com.example.test.MainActivity
import com.example.test.utils.ProgressDialogUtil
import android.app.ProgressDialog
import com.example.test.request.RequestLogin
import okhttp3.ResponseBody
import android.widget.Toast
import android.text.TextWatcher
import android.text.Editable
import android.text.method.PasswordTransformationMethod
import com.example.test.LoginActivity
import android.content.DialogInterface
import android.os.Handler
import android.view.View
import com.example.test.RegisterInfluencerActivity
import com.example.test.RegisterBrandActivity
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.lang.Exception

class LoginActivity : AppCompatActivity() {
    private var input_username: EditText? = null
    private var input_password: EditText? = null
    private var clearEmailLogin: FrameLayout? = null
    private var show: ImageButton? = null
    private var hide: ImageButton? = null
    private var apiInterface: APIInterface? = null
    private var parent_view: View? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        parent_view = findViewById(android.R.id.content)
        input_username = findViewById(R.id.input_username)
        input_password = findViewById(R.id.input_password)
        clearEmailLogin = findViewById(R.id.clearEmailLogin)
        show = findViewById(R.id.show)
        hide = findViewById(R.id.hide)
        input_username?.setFocusable(true)
        findViewById<View>(R.id.btnLogin).setOnClickListener { submitLogin() }
        findViewById<View>(R.id.sign_up).setOnClickListener { showSingleChoiceDialog() }
        apiInterface = APIClient.getClient(applicationContext).create(
            APIInterface::class.java
        )
        val userPicker = ProfileUtil().getUser(applicationContext)
        if (userPicker != null) {
            val intent = Intent(applicationContext, MainActivity::class.java)
            startActivity(intent)
            finish()
        }
        initProperty()
    }

    private fun submitLogin() {
        val email = input_username!!.text.toString().trim { it <= ' ' }
        val password = input_password!!.text.toString().trim { it <= ' ' }
        if (email == "") {
            input_username!!.error = "please input username"
            return
        }
        if (password == "") {
            input_password!!.error = "please input password"
            return
        }
        try {
            val progressDialogUtil =
                ProgressDialogUtil(this@LoginActivity, ProgressDialog.STYLE_SPINNER, false)
            progressDialogUtil.show()
            val data = RequestLogin(email, password)
            val callLogin = apiInterface!!.doLogin(data)
            callLogin.enqueue(object : Callback<ResponseBody> {
                override fun onResponse(
                    call: Call<ResponseBody>,
                    response: Response<ResponseBody>
                ) {
                    if (response.isSuccessful) {
                        val user = response.body().toString()
                        ProfileUtil().save(
                            applicationContext,
                            input_username!!.text.toString(),
                            "Brand",
                            "test@gmail.com"
                        )
                        val intent = Intent(applicationContext, MainActivity::class.java)
                        startActivity(intent)
                        finish()
                        //
                    } else {
                        try {
                            Toast.makeText(
                                this@LoginActivity,
                                response.body().toString(),
                                Toast.LENGTH_SHORT
                            ).show()
                        } catch (e: Exception) {
                            e.message
                        }
                    }
                    progressDialogUtil.dismiss()
                }

                override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
                    Toast.makeText(this@LoginActivity, "Invalid connection", Toast.LENGTH_SHORT)
                        .show()
                    progressDialogUtil.dismiss()
                }
            })
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    private fun initProperty() {
        input_username!!.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable) {
                if (input_username!!.text.toString().length == 0) {
                    input_username!!.visibility = View.VISIBLE
                    clearEmailLogin!!.visibility = View.INVISIBLE
                } else {
                    input_username!!.error = null
                }
            }

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                clearEmailLogin!!.visibility = View.VISIBLE
            }
        })
        input_password!!.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable) {
                if (input_password!!.text.toString().length == 0) {
                    show!!.visibility = View.INVISIBLE
                    hide!!.visibility = View.INVISIBLE
                } else if (hide!!.visibility == View.VISIBLE) {
                    show!!.visibility = View.INVISIBLE
                } else {
                    show!!.visibility = View.VISIBLE
                }
            }

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
        })
        show!!.setOnClickListener {
            input_password!!.transformationMethod = null
            show!!.visibility = View.INVISIBLE
            hide!!.visibility = View.VISIBLE
        }
        hide!!.setOnClickListener {
            input_password!!.transformationMethod = PasswordTransformationMethod()
            show!!.visibility = View.VISIBLE
            hide!!.visibility = View.INVISIBLE
        }
        clearEmailLogin!!.setOnClickListener {
            input_username!!.setText("")
            input_username!!.visibility = View.VISIBLE
        }
    }

    var doubleBackToExitPressedOnce = false
    override fun onBackPressed() {
        //Checking for fragment count on backstack
        if (supportFragmentManager.backStackEntryCount > 0) {
            supportFragmentManager.popBackStack()
        } else if (!doubleBackToExitPressedOnce) {
            doubleBackToExitPressedOnce = true
            Toast.makeText(this, "Press BACK again to exit.", Toast.LENGTH_SHORT).show()
            Handler().postDelayed({ doubleBackToExitPressedOnce = false }, 2000)
        } else {
            super.onBackPressed()
            return
        }
    }

    private var single_choice_selected: String? = null
    private fun showSingleChoiceDialog() {
        single_choice_selected = PILIHAN[0]
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Model User")
        builder.setSingleChoiceItems(PILIHAN, 0) { dialogInterface, i ->
            single_choice_selected = PILIHAN[i]
        }
        builder.setPositiveButton("OK") { dialogInterface, i ->
            if (single_choice_selected == "Influencer") {
                val intent = Intent(this@LoginActivity, RegisterInfluencerActivity::class.java)
                startActivity(intent)
            } else {
                val intent = Intent(this@LoginActivity, RegisterBrandActivity::class.java)
                startActivity(intent)
            }
            finish()
            //                Snackbar.make(parent_view, "selected : " + single_choice_selected, Snackbar.LENGTH_SHORT).show();
        }
        builder.setNegativeButton("CANCEL", null)
        builder.show()
    }

    companion object {
        private val PILIHAN = arrayOf(
            "Influencer", "Brand"
        )
    }
}